src/day16.scala

```diff
@@ -0,0 +1,163 @@
+// using scala 3.1.0
+
+package day16
+
+import scala.util.Using
+import scala.io.Source
+import scala.annotation.tailrec
+
+@main def part1(): Unit =
+  println(s"The solution is ${part1(readInput())}")
+
+@main def part2(): Unit =
+  println(s"The solution is ${part2(readInput())}")
+
+def readInput(): String =
+  Using.resource(Source.fromFile("input/day16"))(_.mkString)
+
+val hexadecimalMapping =
+  Map(
+    '0' -> "0000",
+    '1' -> "0001",
+    '2' -> "0010",
+    '3' -> "0011",
+    '4' -> "0100",
+    '5' -> "0101",
+    '6' -> "0110",
+    '7' -> "0111",
+    '8' -> "1000",
+    '9' -> "1001",
+    'A' -> "1010",
+    'B' -> "1011",
+    'C' -> "1100",
+    'D' -> "1101",
+    'E' -> "1110",
+    'F' -> "1111"
+  )
+
+/*
+ * Structures for all possible operators
+ */
+enum Packet(version: Int, typeId: Int):
+  case Sum(version: Int, exprs: List[Packet]) extends Packet(version, 0)
+  case Product(version: Int, exprs: List[Packet]) extends Packet(version, 1)
+  case Minimum(version: Int, exprs: List[Packet]) extends Packet(version, 2)
+  case Maximum(version: Int, exprs: List[Packet]) extends Packet(version, 3)
+  case Literal(version: Int, literalValue: Long) extends Packet(version, 4)
+  case GreaterThan(version: Int, lhs: Packet, rhs: Packet) extends Packet(version, 5)
+  case LesserThan(version: Int, lhs: Packet, rhs: Packet) extends Packet(version, 6)
+  case Equals(version: Int, lhs: Packet, rhs: Packet) extends Packet(version, 7)
+
+  def versionSum: Int =
+    this match
+      case Sum(version, exprs)            => version + exprs.map(_.versionSum).sum
+      case Product(version, exprs)        => version + exprs.map(_.versionSum).sum
+      case Minimum(version, exprs)        => version + exprs.map(_.versionSum).sum
+      case Maximum(version, exprs)        => version + exprs.map(_.versionSum).sum
```

Comment:
I think you can use `|` in pattern matches:
```
case x: Sum | Product | Minimum | Maximum
```
and 
```
case y: GreaterThabn | LesserThan | Equals
```

Reply:
That doesn't work unfortunately:
```
[error] ./src/day16.scala:53:12: Illegal variable x in pattern alternative
[error]       case x : Sum | Product | Minimum| Maximum => x.version + x.exprs.map(_.versionSum).sum
[error]            ^^^^^^^
```